mongodb: Depends on libpcap for Linuxbrew

timotheecour opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Linuxbrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Linuxbrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

fix https://github.com/Linuxbrew/homebrew-core/pull/235

by adding `depends_on "homebrew/dupes/libpcap"`

path/brew/Cellar/mongodb/3.4.1: 16 files, 2.0G, built in 7 minutes 26 seconds

@sjackman 

EDIT: looks like the issue this fixes has been turned into a PR (also pending on travis failing):  https://github.com/Linuxbrew/homebrew-core/pull/235

timotheecour commented:
@sjackman PTAL; I reproduced the issue on `circle` by using a docker instance of ubuntu:
```
docker run -i -t phusion/baseimage /bin/bash
apt-get -qq update && apt-get -qq -y install build-essential curl git python-setuptools ruby
useradd -ms /bin/bash linuxbrew && su linuxbrew
cd ~
git clone https://github.com/Linuxbrew/brew.git ~/.linuxbrew
export PATH="$HOME/.linuxbrew/bin:$PATH"
brew install mongodb
```

(feel free to suggest improvements in the above commands)
then I hacked the formula until it worked

NOTE: 
* it's very slow to build from source. would be nice to have a bottle soon.
* is compilation really running multi-threaded? not sure it does

NOTE: just saw circleci fail after 1 hour but ... I don't have the logs (cf https://github.com/Linuxbrew/homebrew-core/issues/1581), it could be time-out

sjackman commented:
There's a home for `Dockerfile` for Linuxbrew at https://github.com/linuxbrew/docker
There's a Travis image, and a Circle image would be useful.

sjackman commented:
```
Checking if C compiler is GCC 5.3.0 or newer...no
Checking if C++ compiler is GCC 5.3.0 or newer...no
ERROR: Refusing to build with compiler that does not meet requirements
See /tmp/mongodb-20170115-38567-1bj7ora/mongodb-src-r3.4.1/build/scons/config.log for details
```
Well that's interesting. We'll have to add a `fails_with :gcc => "4.8"`

timotheecour commented:
Sorry what do you recommend I do here? add `fails_with :gcc => "4.8"` in the formula?

or (reading http://www.rubydoc.info/github/Homebrew/brew/Formula.fails_with) this?

fails_with :gcc => '4.8' do
  version '5.3.0'
end

sjackman commented:
Add to the `mongodb` formula:
```ruby
fails_with :gcc => "4.8"
```

sjackman commented on Formula/mongodb.rb:
```diff
@@ -49,7 +50,14 @@ def install
+        ENV["CGO_CPPFLAGS"] = "-I " + Formula["libpcap"].opt_include
+        ENV["CGO_LDFLAGS"] = "-L " + Formula["libpcap"].opt_lib
+        # TODO: improve this, eg by using *args instead of ssl, etc
+        system "bash -c 'source ./set_gopath.sh && bash ./build.sh ssl'"
```
What does `./set_gopath.sh` do? `bash -c` shouldn't be necessary. Try something like (untested)
```ruby
system "source ./set_gopath.sh && ./build.sh #{args.join " "}"
```

timotheecour replied:
Fixed. I removed `./set_gopath.sh` (was in upstream instructions) ; without bash I get:
` ./build.sh: 10: ./build.sh: Bad substitution` or `sh: 1: source: not found.`

timotheecour commented:
PTAL

sjackman commented:
Travis
```
$ travis_wait brew test-bot
Still running (20 of 20): brew test-bot
Timeout (20 minutes) reached. Terminating "brew test-bot"
The command brew test-bot exited with 137.
Log:
cat: travis_wait_1663.log: No such file or directory
The command "travis_wait brew test-bot" exited with 137.
Done. Your build exited with 1.
```
https://travis-ci.org/Linuxbrew/homebrew-core/builds/192631215#L524

sjackman commented:
Circle
```
==> Downloading https://fastdl.mongodb.org/src/mongodb-src-r3.4.1.tar.gz
Already downloaded: /home/ubuntu/.cache/Homebrew/mongodb-3.4.1.tar.gz
==> Verifying mongodb-3.4.1.tar.gz checksum
tar xzf /home/ubuntu/.cache/Homebrew/mongodb-3.4.1.tar.gz
Error: The selected compiler doesn't support C++11: c_compiler
==> FAILED
```
Grumble. This is a bug in `brew` that I introduced in https://github.com/Linuxbrew/brew/commit/b04694b339daa388b267e075012b3123d70d8a1b to allow Linuxbrew to fall back to `/usr/bin/cc` as a compiler to bootstrap an installation of Linuxbrew, but the compiler selection logic doesn't know that it is in fact GCC 4.8, and it's going to fail on this formula.

Could you please try adding
```ruby
depends_on "gcc" # Fails with GCC < 5.3
```
There may be a more general fix, but that's worth trying for now.

timotheecour commented:
> We can increase the timeout parameter of travis_wait.
sure

> I don't know why the log travis_wait_1663.log is missing.

I think the explanation is here: https://github.com/bioconda/bioconda-recipes/issues/3130#issuecomment-265735784

>> The actual problem is in travis_wait, which does the stdout buffering: travis-ci/travis-ci#5716

Could you try to modify the tester script with: 
`brew install homebrew/dupes/expect`
and then use `unbuffer travis_wait` ?
CF https://github.com/travis-ci/travis-ci/issues/5716#issuecomment-273388518

timotheecour commented:
> Could you please try adding

done, PTAL

sjackman commented on Formula/mongodb.rb:
```diff
@@ -49,7 +52,14 @@ def install
+      else
+        ENV["CGO_CPPFLAGS"] = "-I " + Formula["libpcap"].opt_include
+        ENV["CGO_LDFLAGS"] = "-L " + Formula["libpcap"].opt_lib
+        # NOTE: using bash to avoid error: ./build.sh: 10: ./build.sh: Bad substitution
```
If line 10 of the script is a bashism (not portable shell), can you report that bug upstream? Either they should replace the bashism with portable shell, or keep the bashism and change the shebang line to `#!/bin/bash`.

timotheecour replied:
https://github.com/mongodb/mongo-tools/issues/87

sjackman commented on Formula/mongodb.rb:
```diff
@@ -49,7 +52,14 @@ def install
+        ENV["CGO_CPPFLAGS"] = "-I " + Formula["libpcap"].opt_include
+        ENV["CGO_LDFLAGS"] = "-L " + Formula["libpcap"].opt_lib
+        # NOTE: using bash to avoid error: ./build.sh: 10: ./build.sh: Bad substitution
+        system("bash ./build.sh #{args.join ' '}")
```
Try instead:
```sh
system "bash", "./build.sh", *args
```

sjackman commented on Formula/mongodb.rb:
```diff
@@ -15,11 +15,14 @@ class Mongodb < Formula
   option "with-boost", "Compile using installed boost, not the version shipped with mongodb"
   option "with-sasl", "Compile with SASL support"
 
+  depends_on "gcc" # Fails with GCC < 5.3
```
Please add `if OS.linux?`
```ruby
depends_on "gcc" if OS.linux? # Fails with GCC < 5.3
```

sjackman commented:
Circle
`Your build has exceeded the memory limit of 4G on 1 container.`
Try cherry-picking this commit to address this issue:
https://github.com/Linuxbrew/homebrew-core/pull/235/commits/8c09524494e1cf0c88186b10367ed128f453f0af

timotheecour commented:
@sjackman PTAL addressed all comments
> Try cherry-picking this commit to address this issue 8c09524


```
git cherry-pick 8c09524494e1cf0c88186b10367ed128f453f0af
fatal: bad object 8c09524494e1cf0c88186b10367ed128f453f0af
```
so i manually cherry picked

sjackman commented:
TravisCI
```
Building bsondump...
# pkg-config --cflags openssl
pkg-config: exec: "pkg-config": executable file not found in $PATH
```
https://travis-ci.org/Linuxbrew/homebrew-core/builds/200776192#L2102

Add
```ruby
depends_on "pkg-config" => :build unless OS.mac?
```
squash and rebase onto master.

Thanks for sticking with this PR. It's been a tricky one.

timotheecour commented:
@sjackman PTAL. I  added pkg-config, and moved all the linux specific changes into:
```
 if OS.linux?
    depends_on "gcc" # Fails with GCC < 5.3
    depends_on "homebrew/dupes/libpcap" => :build
    depends_on "pkg-config" => :build
  end
```

then squashed and cherry-picked in master. NOTE: rebase caused lots of issues (maybe u r rebasing master urself?) so instead i cherry-picked

sjackman commented:
TravisCI
`The job exceeded the maximum time limit for jobs, and has been terminated.`
https://travis-ci.org/Linuxbrew/homebrew-core/builds/200961868#L2200

sjackman commented:
The `mongodb` bottles is 705 MB. Is that expected? I'm hesitant to merge such a large bottle. I could merge it without the bottle.

sjackman commented:
Brew bottles are not stripped of debugging symbols by default. I've seen that programs that use a lot of C++ templates can result in very large executables with debugging symbols. Perhaps that's an issue here.

sjackman commented:
```
668M	mongod
654M	mongoperf
318M	mongos
269M	mongo
```
```sh
strip mongod mongoperf mongos mongo
```
```
 44M	mongod
 43M	mongoperf
 24M	mongos
 23M	mongo
```

Yep. That's it alright.

@Linuxbrew/core Normally debugging symbols aren't this bug, but it does come up fairly often for C++ programs. Homebrew used to strip executables by default, but stopped doing that, and I don't know the reasoning. Should we strip executables by default? Or perhaps just on a case-by-case basis?

@timotheecour For now, can you add
```ruby
# Reduce the bottle size from 700 MB.
system "strip", Dir[bin/"*"] unless OS.mac?
```

rwhogg commented:
> Should we strip executables by default?

I'd like to know more about Homebrew's rationale before making a general decision.

What might be neat is something similar to Mac's Mach-O system, whereby after installation and before linking we strip all executables on an opt-in basis. (Just throwing an idea out there!)

> Or perhaps just on a case-by-case basis?

If the bottle is relocatable and very large (> 500 MB? Just a random number, no reason in particular for it) then for now I think stripping it is worthwhile. The reason is relocatable bottles will contribute to peoples' disk quotas; non-relocatable ones will only ever get installed by the linuxbrew user so it's less important. Of course, huge downloads suck no matter which way you look at it.

sjackman commented:
> What might be neat is something similar to Mac's Mach-O system, whereby after installation and before linking we strip all executables on an opt-in basis. (Just throwing an idea out there!)

A `brew strip` command that strips the specified formulae or all formulae with no argument (or perhaps `--all`) could be quite useful.

> If the bottle is relocatable and very large (> 500 MB? Just a random number, no reason in particular for it) then for now I think stripping it is worthwhile. The reason is relocatable bottles will contribute to peoples' disk quotas; non-relocatable ones will only ever get installed by the linuxbrew user so it's less important. Of course, huge downloads suck no matter which way you look at it.

I'd set the rough threshold lower to say 200 MB. I'd also like to get more Linuxbrew installations into the default prefix to make more use of bottles.

sjackman commented:
@adamv What was the reasoning behind "Stop stripping binary files" https://github.com/Homebrew/legacy-homebrew/commit/54618074a174026f1231d9a4bd7088d76ab0ee01 ? I wasn't able to find the PR associated with this commit. Some bottles in Linuxbrew like `mongodb` are 700 MB without stripping and ~150 MB with stripping. It's not typical, but the debugging symbols of some C++ programs that use templates a lot can take up a lot of space.

sjackman commented:
Debugging symbols seem to take up less space with `clang`. I wonder if that's also true of `gcc-6.` We're using `gcc-5` currently.

adamv commented:
It was a long time ago, but if I remember it was something like there were some software packages that don't work when symbols were stripped

sjackman commented:
@timotheecour Let's strip the executables to reduce the file size. Please add to the end of `install`. With that change we can finally merge this long-lived PR.
```ruby
system "strip", *Dir[bin/"*"].select { |f| Pathname.new(f).elf? }
```
Please also rebase onto `master`.

maxim-belkin commented:
@sjackman stripping affects ability to [debug](https://felix.abecassis.me/2012/08/gdb-debugging-stripped-binaries/)

sjackman commented:
Of course, but stripping reduces the bottle size from 700 MB to a tenth of that.

maxim-belkin commented:
rebased and force-pushed

sjackman commented:
I've created PR https://github.com/Linuxbrew/homebrew-core/pull/2941 to pick up where this PR left off.
Thanks for all your work on this issue, Timothee.
Superseded by PR https://github.com/Linuxbrew/homebrew-core/pull/2941.
